feat: Manage Nervos DAO with multisig address

devchenyan opened the pull request:
https://github.com/Magickbase/neuron-public-issues/issues/425

https://github.com/user-attachments/assets/a6b26e0e-bbd4-404f-8ce9-d4c039e1ac6a

Danie0918 commented:
@homura @yanguoyu Please have a review.

devchenyan commented on packages/neuron-wallet/src/services/tx/transaction-generator.ts:
```diff
@@ -541,9 +566,20 @@ export class TransactionGenerator {
     fee: string = '0',
-    feeRate: string = '0'
+    feeRate: string = '0',
+    lockClass = {
+      lockArgs: [''],
+      codeHash: SystemScriptInfo.SECP_CODE_HASH,
+      hashType: ScriptHashType.Type,
+    },
+    multisigConfig?: MultisigConfigModel
   ): Promise<Transaction> => {
-    const secpCellDep = await SystemScriptInfo.getInstance().getSecpCellDep()
+    let cellDep: CellDep
+    if (lockClass.codeHash === SystemScriptInfo.MULTI_SIGN_CODE_HASH) {
+      cellDep = await SystemScriptInfo.getInstance().getMultiSignCellDep()
+    } else {
+      cellDep = await SystemScriptInfo.getInstance().getSecpCellDep()
+    }
```
The `lockClass` is still needed in subsequent processing

yanguoyu commented:
I guess it will be better to back to the multisig address table after the send success dialog opens and closes
![image](https://github.com/user-attachments/assets/707f9961-457a-4834-a83d-e312e008375a)

yanguoyu commented on packages/neuron-ui/src/components/DepositDialog/hooks.ts:
```diff
@@ -239,22 +268,37 @@ export const useBalanceReserved = () => {
+        type: AppActions.RequestPassword,
+        payload: {
+          walletID: wallet.id,
+          actionType: multisigConfig.m === 1 ? 'send-from-multisig-need-one' : 'send-from-multisig',
```
This can be verified using the `getMultisigSignStatus` result named `canBroadcastAfterSign`.

Danie0918 commented:
@homura @yanguoyu Please have a review.

silySuper commented:
/package

silySuper commented:
[Dao deposit]
1.When deposit successfully,history page does not show records
<img width="934" alt="截屏2025-02-07 17 35 04" src="https://github.com/user-attachments/assets/8258f19a-3274-4e4c-a896-08b187f0faac" />

history page:no 17:28 and 17:24 records
<img width="1157" alt="截屏2025-02-07 17 35 33" src="https://github.com/user-attachments/assets/58a64d93-3382-4994-8658-976bf3d3e8c6" />

2.When deposit successfully balance does not change .original balance is 900ckb,after two deposits(102CKB every record) it is 899 ckb.

<img width="920" alt="截屏2025-02-07 17 50 27" src="https://github.com/user-attachments/assets/efad66f0-8a6a-4ecd-a738-d2d156b2405d" />

<img width="791" alt="截屏2025-02-07 17 50 37" src="https://github.com/user-attachments/assets/38411efa-eeb5-4ad5-8b46-61ca3f750480" />

3.Deposit need add success tip after deposit successfully

[dao withdraw]
4.view transaction detail jump to undefined page,status is before withdraw.

https://github.com/user-attachments/assets/86aed094-bd04-42e1-be4a-4c6fc7eeba3d

yanguoyu commented on packages/neuron-wallet/src/services/cells.ts:
```diff
@@ -223,13 +223,29 @@ export default class CellsService {
       .leftJoinAndSelect('output.transaction', 'tx')
       .where(
-        `
+        lockArgs
+          ? `
+        output.daoData IS NOT NULL AND
+        output.lockArgs = :lockArgs AND
+        (
+          output.status = :liveStatus OR
+          output.status = :sentStatus OR
+          tx.status = :failedStatus OR
+          (
+            (
+              output.status = :deadStatus OR
+              output.status = :pendingStatus
+            ) AND
+            output.depositTxHash is not null
+          )
+        )`
+          : `
```
```suggestion
        `
        output.daoData IS NOT NULL AND
        (
          output.status = :liveStatus OR
          output.status = :sentStatus OR
          tx.status = :failedStatus OR
          (
            (
              output.status = :deadStatus OR
              output.status = :pendingStatus
            ) AND
            output.depositTxHash is not null
          )
        ) AND` + lockArgs
          ? `output.lockArgs = :lockArgs`
          : `output.lockArgs in (
          SELECT publicKeyInBlake160
          FROM hd_public_key_info
          WHERE walletId = :walletId
        )`,
```

silySuper commented:
5.[dao deposit]
In deposit window,deposit amount does not change from 0 to max ,change from 0 ->bigger than max ->max

https://github.com/user-attachments/assets/ffe92d7d-30bb-4a29-819b-81205a4c46f0


6.When enter wrong password ,it shows wallet is not correct,maybe shows password is wrong is better.
<img width="718" alt="截屏2025-02-08 10 15 44" src="https://github.com/user-attachments/assets/e7676b73-08b6-4e25-b5f6-c5dad5cddb63" />

7.Data is highlight in concise data,does it have other meaning?
<img width="840" alt="截屏2025-02-08 10 18 56" src="https://github.com/user-attachments/assets/1d33f003-b63e-4b36-bd96-562cbff24966" />

8.When no more than 102CKB,tip does not show ,and can also proceed.After enter right password it shows wrong wallet.
<img width="672" alt="截屏2025-02-08 11 19 38" src="https://github.com/user-attachments/assets/56ecdec5-30fe-449c-8279-b15e109cf7d5" />


https://github.com/user-attachments/assets/ec8356a9-c52f-47b4-bf75-cfcbc6be1558

9.[dao withdraw]
In dark mode,font is black.In expect it should be white.
<img width="923" alt="截屏2025-02-08 11 24 29" src="https://github.com/user-attachments/assets/db466450-d0c1-46a8-a220-9a1fe367aa10" />

<img width="752" alt="截屏2025-02-08 11 27 59" src="https://github.com/user-attachments/assets/d1212bf2-2794-48c0-b984-03be636c3a8e" />

10.withdraw UI is different from UI document.
 UI document:
<img width="694" alt="截屏2025-02-08 11 31 59" src="https://github.com/user-attachments/assets/2d24e12c-5496-4ab4-a23b-79840461220a" />

actual UI:
<img width="694" alt="截屏2025-02-08 11 31 59" src="https://github.com/user-attachments/assets/f720d188-70f6-4136-a16c-0bcedfbffeea" />

Danie0918 commented:
@yanguoyu @silySuper Please have a review.

silySuper commented:
/package

silySuper commented:
1.same as before
2.same as before
3.same as before
4.same as before
5.same as before
6.same as before
7..same as before
8.same as before
9.verified
10.same as before

please check whether code is submitted  correctly. @devchenyan

devchenyan commented:
1. Transaction history will not display transactions of multisig addresses
2. Since the outputs are also the multisig address itself, the balance remains the same
3. Fixed
4. Fixed
5.  Needs to be modified together with NervosDAO
6. General logic for verifying multisig address signatures
7. Data style when `data !== '0x'`
8. Needs to be modified together with NervosDAO
9. Fixed
10. The UI document matches the actual UI

Keith-CY commented:
/package

silySuper commented:
> 1. Transaction history will not display transactions of multisig addresses
> 2. Since the outputs are also the multisig address itself, the balance remains the same
> 3. Fixed
> 4. Fixed
> 5. Needs to be modified together with NervosDAO
> 6. General logic for verifying multisig address signatures
> 7. Data style when `data !== '0x'`
> 8. Needs to be modified together with NervosDAO
> 9. Fixed
> 10. The UI document matches the actual UI

2.This may cause problem when no balance to deposit but show enough balance.
5.Nervos dao is normal,this is video:

https://github.com/user-attachments/assets/3201fb21-346a-496d-9b4a-27e142cbe04e

6.Approve shows like this when password is wrong:
<img width="868" alt="截屏2025-02-11 12 23 40" src="https://github.com/user-attachments/assets/7a8fcf6f-eed6-4e0c-8a89-a3ceefd79f88" />

in Nervos Dao shows like this when password is wrong:

<img width="843" alt="截屏2025-02-11 12 24 45" src="https://github.com/user-attachments/assets/fe5487d0-aef1-43a3-b86a-1e60b61d9e41" />

8.Nervos Dao show tip when balance is not enough and proceed button is grey:
<img width="914" alt="截屏2025-02-11 12 34 35" src="https://github.com/user-attachments/assets/9008457c-9b83-4171-90a8-95503b9aa7ed" />

10.This is UI document,before upload wrong.(issue before amend also)
![image](https://github.com/user-attachments/assets/acc87324-a165-4fc8-b48d-10030f18fe8b)

silySuper commented:
/package

silySuper commented:
2.verified
3.verified
4.verified
5.verified
6.It is my right wallet,I found that  in approve function if I click sign&export it shows not correct wallet,but if I click confirm it shows incorrect password,it is a history bug.I will open another link issue to this bug.
8.verified

thetdy commented:
when using a ledger with multisig it only seems to work sending ckb but not for the DAO deposits. 


![image](https://github.com/user-attachments/assets/46034a4e-684a-4643-aa0c-e36dcae045fc)

Keith-CY commented:
> when using a ledger with multisig it only seems to work sending ckb but not for the DAO deposits.
> 
> tested on linux and windows.
> 
> ![image](https://private-user-images.githubusercontent.com/51980251/421134047-46034a4e-684a-4643-aa0c-e36dcae045fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE2NzI1MTksIm5iZiI6MTc0MTY3MjIxOSwicGF0aCI6Ii81MTk4MDI1MS80MjExMzQwNDctNDYwMzRhNGUtNjg0YS00NjQzLWFhMGMtZTM2ZGNhZTA0NWZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzExVDA1NTAxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5MGIwMDY2ZTc0MTU1NGRmYjZmYTNjNDNhYTJlMWQ2ZWQ2ODdhNjk0MWE4MTUwMzA5MGQ3YTc0MDI3MmQ3ZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gj01Kzj9IkAajui75v-x9LJiNTJHoo7NqLdeu8ly1kM)

@Danie0918 please check this

Danie0918 commented:
> when using a ledger with multisig it only seems to work sending ckb but not for the DAO deposits.
> 
> tested on linux and windows.
> 
> ![image](https://private-user-images.githubusercontent.com/51980251/421134047-46034a4e-684a-4643-aa0c-e36dcae045fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3NTA5NDEsIm5iZiI6MTc0MTc1MDY0MSwicGF0aCI6Ii81MTk4MDI1MS80MjExMzQwNDctNDYwMzRhNGUtNjg0YS00NjQzLWFhMGMtZTM2ZGNhZTA0NWZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEyVDAzMzcyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiMGZmY2UyNmE4NmNhNjU0NDc0NjQwMDI4MzRhYTVkMDIyMWYxZTU2ZWVmOGI2ZmUzNTRlZjgzMTg4NDhjMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-VS_7yxqsaYCkRi5KqAhtJ2R5L2AB_4J2_aVQXbqTlU)

Thanks for your feedback! We have confirmed that Ledger does not support DAO deposits and withdrawals for multisig addresses and will address this issue.https://github.com/Magickbase/neuron-public-issues/issues/446

However, we were unable to reproduce the error shown in your screenshot. Could you please provide the steps to reproduce the issue or export the logs to help us understand the situation? (Menu > Help > Export Debug Information)
